File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

function first_then_lower_case(string_table, predicate)

  for _, string in pairs(string_table) do
    if predicate(string) then
      return string.lower(string)
    end
  end
  return nil
end

function powers_generator(base, limit)
  local power = 1
  return coroutine.create(function ()
    while power <= limit do
      coroutine.yield(power)
      power = power * base
    end
  end)
end

function say(word)
  if word == nil then
    return ""
  end
  return function(next)
    if next == nil then
      return word
    else
      return say(word .. " " .. next)
    end
  end
end

function meaningful_line_count(fileName)
  local file, _ = io.open(fileName, "r")

  if not file then
      error(string.format("No such file", fileName))
  end

  local lineCount = 0

  for line in file:lines() do
      local strippedLine = line:gsub("^%s*(.-)%s*$", "%1")
      if strippedLine ~= "" and not strippedLine:match("^%s*#") then
          lineCount = lineCount + 1
      end
  end

  file:close()

  return lineCount
end


Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
    local self = setmetatable({}, Quaternion)
    self.a = a or 0
    self.b = b or 0
    self.c = c or 0
    self.d = d or 0
    return self
end

function Quaternion:__tostring()
  local parts = {}

    if self.a ~= 0 then
      table.insert(parts, tostring(self.a))
    end

    if self.b ~= 0 then
      if self.b == 1 then
        table.insert(parts, "+i")
      elseif self.b == -1 then
        table.insert(parts, "-i")
      else
        if self.b > 0 and #parts > 0 then
          table.insert(parts, "+" .. tostring(self.b) .. "i")
        else
          table.insert(parts, tostring(self.b) .. "i")
        end
      end
    end

    if self.c ~= 0 then
      if self.c == 1 then
        table.insert(parts, "+j")
      elseif self.c == -1 then
        table.insert(parts, "-j")
      else
        if self.c > 0 and #parts > 0 then
          table.insert(parts, "+" .. tostring(self.c) .. "j")
        else
          table.insert(parts, tostring(self.c) .. "j")
        end
      end
    end

    if self.d ~= 0 then
      if self.d == 1 then
        table.insert(parts, "+k")
      elseif self.d == -1 then
        table.insert(parts, "-k")
      else
        if self.d > 0 and #parts > 0 then
          table.insert(parts, "+" .. tostring(self.d) .. "k")
        else
          table.insert(parts, tostring(self.d) .. "k")
        end
      end
    end

  local result = table.concat(parts)

  if result == "" then
    return "0"
  end
  if result:sub(1, 1) == "+" then
    result = result:sub(2)
  end

  return result
end

function Quaternion:__add(other)
  if getmetatable(other) == Quaternion then
    return Quaternion.new(
      self.a + other.a,
      self.b + other.b,
      self.c + other.c,
      self.d + other.d
    )
  else
    return nil
  end
end

function Quaternion:__mul(other)
  if getmetatable(other) == Quaternion then
    local a1, b1, c1, d1 = self.a, self.b, self.c, self.d
    local a2, b2, c2, d2 = other.a, other.b, other.c, other.d

    return Quaternion.new(
      a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
      a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
      a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
      a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
    )
  else
    return nil
  end
end

function Quaternion:__eq(other)
  if getmetatable(other) == Quaternion then
    return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
  else
    return false
  end
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


def first_then_lower_case(string_list, predicate, /):

    for string in string_list:
        if predicate(string):
            return string.lower()

    return None
    

def powers_generator(*, base, limit):
    power = 1
    while power <= limit:
        yield power
        power = power * base


def say(word=None):
    if word == None:
        return ""
    def next_word(next=None):
        if next == None:
            return word
        return say(word + " " + next)
    return next_word

def meaningful_line_count(file_name, encoding='utf-8'):
    line_count = 0
    with open(file_name, 'r', encoding=encoding) as file:
        for line in file:
            stripped_line = line.strip()
            if stripped_line and not stripped_line.startswith('#'):
                line_count += 1
        return line_count

@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    def __repr__(self):
        """String representation of the quaternion"""
        return f"{self.a} + {self.b}i + {self.c}j + {self.d}k"
    
    def __str__(self):
        parts = []

        if self.a != 0:
            parts.append(f"{self.a}")

        if self.b != 0:
            if self.b == 1:
                parts.append("+i")
            elif self.b == -1:
                parts.append("-i")
            else:
                if self.b > 0 and parts:
                    parts.append(f"+{self.b}i")
                else:
                    parts.append(f"{self.b}i")

        if self.c != 0:
            if self.c == 1:
                parts.append("+j")
            elif self.c == -1:
                parts.append("-j")
            else:
                if self.c > 0 and parts:
                    parts.append(f"+{self.c}j")
                else:
                    parts.append(f"{self.c}j")

        if self.d != 0:
            if self.d == 1:
                parts.append("+k")
            elif self.d == -1:
                parts.append("-k")
            else:
                if self.d > 0 and parts:
                    parts.append(f"+{self.d}k")
                else:
                    parts.append(f"{self.d}k")

        result = "".join(parts)

        if not result:
            return "0"
        if result.startswith("+"):
            result = result[1:]

        return result

    def __add__(self, other):
        """Addition of two quaternions"""
        if isinstance(other, Quaternion):
            return Quaternion(
                self.a + other.a,  
                self.b + other.b,  
                self.c + other.c,  
                self.d + other.d 
            )
        return NotImplemented

    def __mul__(self, other):
        """Multiplication of two quaternions (Hamilton product)"""
        if isinstance(other, Quaternion):
            a1, b1, c1, d1 = self.a, self.b, self.c, self.d
            a2, b2, c2, d2 = other.a, other.b, other.c, other.d

            return Quaternion(
                a1*a2 - b1*b2 - c1*c2 - d1*d2,
                a1*b2 + b1*a2 + c1*d2 - d1*c2,
                a1*c2 - b1*d2 + c1*a2 + d1*b2,
                a1*d2 + b1*c2 - c1*b2 + d1*a2
            )
        return NotImplemented

    def __eq__(self, other):
        """Value-based equality between two quaternions"""
        if isinstance(other, Quaternion):
            return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
        return False

    @property
    def conjugate(self):
        """Return the conjugate of the quaternion"""
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    @property
    def coefficients(self):
        """Return the coefficients (a, b, c, d) as a list"""
        return (self.a, self.b, self.c, self.d)